### PR TITLE
transforms: Canonicalize without the MLContext

### PR DESCRIPTION
**Pull request by math-fehr**
Do not rely on the loaded operations in the canonicalization pass.

This change is necessary to let dialects be loaded only during parsing. An alternative solution would be to load all dialects before doing the rewrites, or to load dialects automatically during imports.

### CODE REVIEW

**Comment by superlopuh** [comment]
I'm not sure I get it, what's the benefit of this? Or rather, what breaks if this change isn't made?

**Comment by math-fehr** [comment]
The main issue is coming from my "lazy dialects" change.

The overall goal is to not load files that do not need to be loaded, since we use a non-trivial amount of time loading Python files. My current solution is to never load dialect files, and instead load Python files dynamically during parsing, or during a pass call. However, that means in particular that we cannot get all dialects without loading all of them anymore, which is becoming costly.

In short, the `MLContext` will contain load the dialects that have been parsed (I'm going to remove most of `MLContext` uses btw). This means that we can't rely on it anymore for that pass (and it is the only pass requiering this). It will definitely have a cost for that pass, but there is probably room of improvement (like walking the pass first, to build the rewritepattern).

**Comment by superlopuh** [comment]
I'm still not sure that I see the issue. If the ops are in a module with a context, then it should be safe to assume that their dialect is loaded in the context, also, no?

**Comment by math-fehr** [comment]
In MLIR, this is done by loading dialects before each pass, and noting in each pass which dialects are dependent.
Since we don't do that, we don't have at all that assumption. Also, I'm not sure we actually want to have that, since besides this pass, we have absolutely no need for that context.

The dialect Python file will be loaded though, but the dialect won't necessary be loaded in the context.

**Comment by superlopuh** [comment]
Ah. I would really like for us to have that mechanism in all our passes, though. I think there's quite a lot of benefit of having the context represent the set of legal operations at a point in the compilation pipeline. I can see a few ways it could be used for optimisation and for raising errors if something weird happens. In any case, I can see how that is scope creep wrt this PR.

**Comment by superlopuh** [comment]
Could we not register the dialects of the ops in the current IR? This would be a smaller change wrt the current mechanism, and should not affect the python import stuff since the dialects must already be in the interpreter if the ops have been constructed

**Comment by math-fehr** [comment]
> Could we not register the dialects of the ops in the current IR? This would be a smaller change wrt the current mechanism, and should not affect the python import stuff since the dialects must already be in the interpreter if the ops have been constructed

But that would be much slower right?
The cost of constructing the RewritePatterns will anyway be larger, because here we don't really do anything.
However, in the previous version, the fastest non-match is `O(pattern_size)`, since we do an `isinstance` for each rewrite pattern we have (as we don't optimize them at all). Here, it would be `O(op_trait_size)`, which scales better?

**Review by superlopuh — APPROVED** [review]
Good point, you've convinced me
